### ISSUE TITLE
Bug: DpadSpanSizeLookup's getSpanSize function is being called infinitely after removing an item

### ISSUE THREAD

**Issue by tyrel-carlson** [rubensousa/DpadRecyclerView#210]
Hi there, hope you are doing well,

I'm using DpadRecycleView to show a grid of movie items with spanSize of 4. They are sectioned by date headers.

So I've faced an issue with DpadSpanSizeLookup's getSpanSize function is being called infinitely after removing an item by an action in another emulator.

When I add a movie to the list from another emulator, it works fine and adds it to the grid view, but for removing an item it will call getSpanSize unlimited with the constant position of 10. No matter which item is removed in which position. It always keeps calling with the position=10.

I've replaced DpadRecyclerView with androidx.widget.RecyclerView and used GridLayoutManager(context, 4) and it works as expected.

here is my code:
```
private fun setupRecyclerView(recycler: DpadRecyclerView) {
        recycler.adapter = adapter
        recycler.setSpanCount(SPAN_COUNT)

        recycler.setSpanSizeLookup(object : DpadSpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                val isHeader = adapter.isSectionHeader(position)
                Log.i("DebugTag", "getSpanSize: $position, isSectionHeader: $isHeader")
                return if (position in 0 until adapter.itemCount && isHeader) SPAN_COUNT else 1
            }
        })

        recycler.attachListController(viewModel.listController)
    }
```

and here is part of the logs:
```
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.661 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
18:42:11.662 DebugTag       I  getSpanSize: 10, isSectionHeader: true
```

I've tried to debug the DpadSpanSizeLookup using breakpoints but they won't trigger in this case. so I used these lines too without any success:
```
spanSizeLookup.setSpanGroupIndexCacheEnabled(false)
        spanSizeLookup.setSpanIndexCacheEnabled(false)
```

**Comment by rubensousa** [comment]
@tyrel-carlson is this similar to your setup? 
![image](https://github.com/rubensousa/DpadRecyclerView/assets/10662096/b0a862b8-f2b4-43d3-8e5c-6743ae0a1445)

I'm not able to reproduce this. Can you give me more details about the contents of your adapter and which position you're removing?

**Comment by rubensousa** [comment]
Here's my current test that is passing:

```kotlin
 @Test
    fun testGridRemovalWithSpanLookupDoesNotCrash() {
        onRecyclerView("Change span size") { recyclerView ->
            recyclerView.setSpanSizeLookup(object : DpadSpanSizeLookup() {
                override fun getSpanSize(position: Int): Int {
                    return if (position % 5 == 0) {
                        recyclerView.getSpanCount()
                    } else {
                        1
                    }
                }
            })
        }
        val oldViewBounds = getRelativeItemViewBounds(position = 0)
        mutateAdapter { adapter ->
            adapter.removeAt(adapter.itemCount - 1)
            adapter.removeAt(adapter.itemCount - 1)
            adapter.removeAt(0)
            adapter.removeAt(0)
        }
        assertFocusAndSelection(0)
        assertItemAtPosition(position = 0, item = 2)

        val newViewBounds = getRelativeItemViewBounds(position = 0)
        assertThat(newViewBounds).isEqualTo(oldViewBounds)
    }
```

I'm trying to replicate your setup, but I need more details

**Comment by rubensousa** [comment]
I manage to reproduce it with this test:

```kotlin
 @Test
    fun testGridRemovalWithSpanLookupDoesNotCrash() {
        onRecyclerView("Change span size") { recyclerView ->
            recyclerView.setSpanSizeLookup(object : DpadSpanSizeLookup() {
                override fun getSpanSize(position: Int): Int {
                    val adapter = recyclerView.adapter as AbstractTestAdapter<*>
                    val item = adapter.getItem(position)
                    return if (item % 9 == 0) {
                        recyclerView.getSpanCount()
                    } else {
                        1
                    }
                }
            })
        }
        KeyEvents.pressDown()
        val oldViewBounds = getRelativeItemViewBounds(position = 1)
        mutateAdapter { adapter ->
            adapter.removeAt(3)
            adapter.removeAt(13)
        }
        assertFocusAndSelection(1)
        assertItemAtPosition(position = 1, item = 1)

        val newViewBounds = getRelativeItemViewBounds(position = 1)
        assertThat(newViewBounds).isEqualTo(oldViewBounds)
    }

```

Will take care of it for the next release

**Comment by tyrel-carlson** [comment]
> @tyrel-carlson is this similar to your setup? ![image](https://private-user-images.githubusercontent.com/10662096/335142463-b0a862b8-f2b4-43d3-8e5c-6743ae0a1445.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwNjQ5NzUsIm5iZiI6MTcxNzA2NDY3NSwicGF0aCI6Ii8xMDY2MjA5Ni8zMzUxNDI0NjMtYjBhODYyYjgtZjJiNC00M2QzLThlNWMtNjc0M2FlMGExNDQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMwVDEwMjQzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU3MWU4Y2M1MzBmZDcwZDBhMDM0NzE0OWFmMGFiYjg2MWU2OTMwMzY4MmE5NWU5Y2ZkMTlkMmYyM2E4ODE0YzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-eC0DQpM4ouOLe848pb1-_uGKtZ__jut8NnAY1tiTiQ)
> 
> I'm not able to reproduce this. Can you give me more details about the contents of your adapter and which position you're removing?

Yeah that looks like my own. My date headers fill the grid but are not focusable themselves.

**Comment by tyrel-carlson** [comment]
> I manage to reproduce it with this test:
> 
> ```kotlin
>  @Test
>     fun testGridRemovalWithSpanLookupDoesNotCrash() {
>         onRecyclerView("Change span size") { recyclerView ->
>             recyclerView.setSpanSizeLookup(object : DpadSpanSizeLookup() {
>                 override fun getSpanSize(position: Int): Int {
>                     val adapter = recyclerView.adapter as AbstractTestAdapter<*>
>                     val item = adapter.getItem(position)
>                     return if (item % 9 == 0) {
>                         recyclerView.getSpanCount()
>                     } else {
>                         1
>                     }
>                 }
>             })
>         }
>         KeyEvents.pressDown()
>         val oldViewBounds = getRelativeItemViewBounds(position = 1)
>         mutateAdapter { adapter ->
>             adapter.removeAt(3)
>             adapter.removeAt(13)
>         }
>         assertFocusAndSelection(1)
>         assertItemAtPosition(position = 1, item = 1)
> 
>         val newViewBounds = getRelativeItemViewBounds(position = 1)
>         assertThat(newViewBounds).isEqualTo(oldViewBounds)
>     }
> ```
> 
> Will take care of it for the next release

Thanks for your consideration. By the way, can you tell me how can I manage to fix this until next version is released?

**Comment by rubensousa** [comment]
As a workaround, `notifyDataSetChanged` instead of individual `notifyItemRemoved` calls should fix it for now. But I plan to release this fix today anyway

**Comment by tyrel-carlson** [comment]
Yeah I noticed too that calling `notifyDataSetChanged` is a fix but it prevents loading of item data in my case, so I prefer to wait for your release, thanks for you great work!

**Comment by rubensousa** [comment]
Released in `1.3.0-alpha03` now
